src/ptc/util/jms.clj

```diff
@@ -24,40 +24,65 @@
-   ::copy   :analysis_version_number     :analysisCloudVersion
-   ::copy   :chip_well_barcode           :chipWellBarcode
-   ::copy   :reported_gender             :gender
-   ::copy   :sample_alias                :sampleAlias
-   ::copy   :sample_lsid                 :sampleLsid
-   ::copy   :call_rate_threshold         :callRateThreshold
-   ::chip   :bead_pool_manifest_file     :beadPoolManifestPath
-   ::chip   :cluster_file                :clusterFilePath
-   ::chip   :extended_chip_manifest_file :chipManifestPath
-   ::chip   :gender_cluster_file         :genderClusterFilePath
-   ::chip   :zcall_thresholds_file       :zCallThresholdsPath
-   ::push   :green_idat_cloud_path       :greenIDatPath
-   ::push   :red_idat_cloud_path         :redIDatPath])
+  ["action" "req'd?" "request notification key"   "JMS key"
+   ::copy   true     :analysis_version_number     :analysisCloudVersion
+   ::copy   true     :chip_well_barcode           :chipWellBarcode
+   ::copy   true     :reported_gender             :gender
+   ::copy   true     :sample_alias                :sampleAlias
+   ::copy   true     :sample_lsid                 :sampleLsid
+   ::copy   true     :call_rate_threshold         :callRateThreshold
+   ::chip   true     :bead_pool_manifest_file     :beadPoolManifestPath
+   ::chip   true     :cluster_file                :clusterFilePath
+   ::chip   true     :extended_chip_manifest_file :chipManifestPath
+   ::chip   false    :gender_cluster_file         :genderClusterFilePath
+   ::chip   false    :zcall_thresholds_file       :zCallThresholdsPath
+   ::push   true     :green_idat_cloud_path       :greenIDatPath
+   ::push   true     :red_idat_cloud_path         :redIDatPath
+   ::param  true     :CHIP_TYPE_NAME              :chipName
+   ::param  true     :CHIP_WELL_BARCODE           :chipWellBarcode
+   ::param  true     :INDIVIDUAL_ALIAS            :collaboratorParticipantId
+   ::param  true     :LAB_BATCH                   :labBatch
+   ::param  true     :PARTICIPANT_ID              :participantId
+   ::param  true     :PRODUCT_FAMILY              :productFamily
+   ::param  true     :PRODUCT_NAME                :productName
+   ::param  true     :PRODUCT_ORDER_ID            :productOrderId
+   ::param  true     :PRODUCT_PART_NUMBER         :productPartNumber
+   ::param  true     :REGULATORY_DESIGNATION      :regulatoryDesignation
+   ::param  true     :RESEARCH_PROJECT_ID         :researchProjectId
+   ::param  false    :SAMPLE_ALIAS                :collaboratorSampleId
+   ::param  true     :SAMPLE_GENDER               :gender
+   ::param  true     :SAMPLE_ID                   :sampleId
+   ::param  true     :SAMPLE_LSID                 :sampleLsid])
+
+(def required-jms-keys
+  "Sort all the keys required to handle a JMS message."
+  (letfn [(required? [[_ reqd? _ jms]] (when reqd? jms))]
+    (->> wfl-keys->jms-keys-table
+         (partition-all 4) rest
+         (keep required?) set sort)))
```

Comment:
Does sorting a set in clojure do anything?
I realise clojure has a `hash-set` and a `sorted-set`, but the doc for `set` doesn't mention anything about what kind it returns (surprise surprise)

Reply:
You're right that sorting a set is useless, but the result is actually a seq.
The result of sort is always a seq, which is true of most core functions.
I sorted so any mention of keys in messages would be sorted.

Reply:
… and the result of `set` is always a `set`.

Reply:
Well that's something at least